### ISSUE TITLE
State CONNECTED sometimes is not returned when wifi is turned off while having mobile internet connection (Android 9)

### ISSUE THREAD

**Issue by PhanVanLinh** [pwittchen/ReactiveNetwork#330]
**Describe the bug**
I use below code for tract network connection in application. It working well on my Android 8 device but it some time not working well on my Android 9 when I try to switch between wifi and mobile internet (4G)
```
ReactiveNetwork
                .observeNetworkConnectivity(context)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    Log.i("ReactiveNetwork", "connect " + it)
                }, {
                    Log.i("ReactiveNetwork", "error " + it)
                })
```

**To Reproduce**
- I open app without wifi and mobile connection  (below log line 1 display) 
- I turn on wifi (below log line 2 display) 
- I turn on mobile (no log display) 
- I turn off wifi (below log line 3 + 4 display) 
- I turn on wifi (no log display) 
- I turn off wifi (below log line 5 display). At this time, my phone still have mobile connection but status CONECT never return (I wait for 2 -> 10 minutes). I tested this many time.

And here is the log
```
2019-05-02 11:02:36.426 24510-24510/ I/ReactiveNetwork: connect Connectivity{state=DISCONNECTED, detailedState=IDLE, type=-1, subType=-1, available=false, failover=false, roaming=false, typeName='NONE', subTypeName='NONE', reason='', extraInfo=''}
2019-05-02 11:02:44.812 24510-24510/ I/ReactiveNetwork: connect Connectivity{state=CONNECTED, detailedState=CONNECTED, type=1, subType=0, available=true, failover=false, roaming=false, typeName='WIFI', subTypeName='', reason='null', extraInfo='null'}
2019-05-02 11:03:01.641 24510-24510/ I/ReactiveNetwork: connect Connectivity{state=DISCONNECTED, detailedState=IDLE, type=-1, subType=-1, available=false, failover=false, roaming=false, typeName='NONE', subTypeName='NONE', reason='', extraInfo=''}
2019-05-02 11:03:01.777 24510-24510/ I/ReactiveNetwork: connect Connectivity{state=CONNECTED, detailedState=CONNECTED, type=0, subType=13, available=true, failover=false, roaming=false, typeName='MOBILE', subTypeName='LTE', reason='connected', extraInfo='diginet'}
2019-05-02 11:03:15.958 24510-24510/ I/ReactiveNetwork: connect Connectivity{state=DISCONNECTED, detailedState=IDLE, type=-1, subType=-1, available=false, failover=false, roaming=false, typeName='NONE', subTypeName='NONE', reason='', extraInfo=''}
```

**Expected behavior**
state `CONNECTED` should return

**Smartphone (please complete the following information):**
 - Device:  Xiaomi A2 (android 9), One Plus 5T(android 9), Huawei Y9 (android 8)
 - OS: Android 9, Android 8
 - Library Version: 3.0.2

Let me know if you need more information. Thank you

**Comment by pwittchen** [comment]
Thanks for reporting this. I'll have a look on that.

--
Piotr Wittchen,
http://wittchen.io

czw., 2 maj 2019, 04:24 użytkownik Phan Van Linh <notifications@github.com>
napisał:

**Comment by mseroczynski** [comment]
Any updates?

**Comment by pwittchen** [comment]
Not yet. It's hard for me to guess what causes the problem when it occurs
"sometimes" only. I think, it may be related to the fact that mobile and
network connection are enabled at the same time. Once WiFi is turned off
and mobile is turned on, phone remains connected. That's why state could
not be returned. I'm not sure if that's really a bug or correct behavior.

--
Piotr Wittchen,
http://wittchen.io

wt., 21 maj 2019, 13:47 użytkownik Michał Seroczyński <
notifications@github.com> napisał:

**Comment by AleksPlekhov** [comment]
+1   I have  the same issue on Pixel XL Android Pie. ExtraInfo  is NULL. Below android 9 it works properly.  Using wifi manager i was able to get required data on android 9, but not using "observeNetworkConnectivity"  https://developer.android.com/guide/topics/connectivity/wifi-scan#kotlin

**Comment by mseroczynski** [comment]
In my app I enforce using OkHttp's cache in requests (by rebulding them with CacheControl.FORCE_CACHE when ReactiveNetwork is saying that there is no connection). I've received numerous bug reports from users saying that they in fact have an internet connection (which is working in other apps) but not in mine. This bug in ReactiveNetwork is an important issue now - it kills connectivity and causes users confusion. Core feature of this lib seems broken. 

@pwittchen Do you have any idea on how to repair it or would you suggest replacing ReactiveNetwork with something else? Thank you for your time.

**Comment by pwittchen** [comment]
I'll have a closer look on that. I'm not sure what may be the source of
this problem. It may be the usage of distinct operator from RxJava inside
library or default Android system behavior or something else. I cannot
check it now because I don't have an access to the computer and I'm writing
from my phone. I'll check it later this week. I see that this issue for
sure is related to the fact that one type of connectivity is enabled while
another is being disabled or the other way around. I also have to verify
how broadcast receivers are propagated by the system and if it's done as
you expect because library just wraps them.

--
Piotr Wittchen,
http://wittchen.io

czw., 23 maj 2019, 16:47 użytkownik Michał Seroczyński <
notifications@github.com> napisał:

**Comment by mseroczynski** [comment]
Always when ReactiveNetwork's last state is NOT_CONNECTED and you have Mobile on & Wi-Fi on turning Wi-fi off fixes the state. Maybe some kind of periodical check would be a reasonable hotfix.

By the way, below is the method I am using:
``` 
  @SuppressLint("CheckResult")
  private fun observeConnectionState() {
    ReactiveNetwork.observeNetworkConnectivity(context)
        .map {
          val isConnected = it.state() == NetworkInfo.State.CONNECTED
          val isWifiOrUnlimited = when (it.type()) {
            TYPE_WIFI,
            TYPE_BLUETOOTH,
            TYPE_ETHERNET -> true
            else -> false
          }

          when (isConnected) {
            true -> {
              if (isWifiOrUnlimited) {
                ConnectionState.WIFI_OR_UNLIMITED_CONNECTED
              } else {
                ConnectionState.MOBILE_OR_LIMITED_CONNECTED
              }
            }
            false -> {
              ConnectionState.NOT_CONNECTED
            }
          }
        }
        .distinctUntilChanged()
        .subscribe { connectionState ->
          i { "connectionState: $connectionState" }
          connectionStateRelay.accept(connectionState)
        }
  }
```

**Comment by pwittchen** [comment]
That's helpful information. Thanks.

--
Piotr Wittchen,
http://wittchen.io

czw., 23 maj 2019, 17:19 użytkownik Michał Seroczyński <
notifications@github.com> napisał:

**Comment by pwittchen** [comment]
Hi guys,

Sorry for the late replies. This is my side project and I have limited time and energy to work on that.
I have initial bug-fix for this issue on the [issue-330](https://github.com/pwittchen/ReactiveNetwork/tree/issue-330) branch. See these commits: 
- https://github.com/pwittchen/ReactiveNetwork/commit/43a3daba7a553fff910f345ff9730a6eb70e1819
- https://github.com/pwittchen/ReactiveNetwork/commit/8526981cb0abdb2c2df43c44f1784d56744cd734

Within these commits, unfortunately I had to introduce the state, so the class became more mutable, but I don't know how to solve this without mutability. I'm saving the last connectivity after its emission and compare it with the current connectivity. If the connectivity TYPE changed, last connectivity state was CONNECTED and current connectivity state is DISCONNECTED (case described in the reproduction steps), then I'm emitting last connectivity instead of the current one (e.g. MOBILE CONNECTED instead of WIFI DISCONNECTED). This is not the perfect solution, but it should prevent false-positive information if the device is connected to at least one network. You can have a look on that change and perform a code review if you'd like to. Before creating a PR, I need to add unit test(s) for that change.

**Edit**:

I've noticed that after this fix, when we change connectity as follows:
1. MOBILE AND WIFI DISCONNECTED
2. MOBILE CONNECTED
3. MOBILE DISCONNECTED

then library indicates MOBILE CONNECTED. This is not correct behavior, so the fix have to be adjusted or changed. Optionally, we can replace this initial solution with additional periodical check as it was already mentioned.

Regards,
Piotr

**Comment by pwittchen** [comment]
Ok, it seems that issue from my previous comment was fixed in https://github.com/pwittchen/ReactiveNetwork/commit/33929437a6bc41d4992c90ec5f064e80d2976029 by adding `isNotIdle` check before propagating last connectivity status. Moreover, I'm propagating current and last connectivity status (in that order) in the case, when type changed and previous status was connected (e.g. connected via MOBILE but disconnected from WIFI). Right now, I need to test this, write unit tests and probably extract new code to the separate method. Any feedback is appreciated!

**Comment by pwittchen** [comment]
#307 bug report is probably related to the same issue as this one.

**Comment by pwittchen** [comment]
This bug is fixed by PR #359.
Bug-fix will be available in the upcoming release.
Sorry that you had to wait for so long for it.
